torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,484 @@
+
+namespace torch {
+namespace dynamo {
+using namespace torch::autograd;
+using c10::SymInt;
+
+static PyObject* wrap_variable_list(const variable_list& inputs) {
+  PyObject* pyinput = PyList_New(inputs.size());
+  for (const auto i : c10::irange(inputs.size())) {
+    PyList_SET_ITEM(pyinput, i, THPVariable_Wrap(inputs[i]));
+  }
+  return pyinput;
+}
+
+static PyObject* wrap_int_list(const std::vector<int64_t>& inputs) {
+  PyObject* pyinput = PyTuple_New(inputs.size());
+  for (const auto i : c10::irange(inputs.size())) {
+    PyTuple_SET_ITEM(pyinput, i, PyLong_FromSsize_t(inputs[i]));
+  }
+  return pyinput;
+}
+
+static PyObject* convert_hook_list(std::vector<c10::SafePyObject>& inputs) {
+  // inplace, consumes the input hooks
+  PyObject* pyinput = PyTuple_New(inputs.size());
+  for (const auto i : c10::irange(inputs.size())) {
+    PyTuple_SET_ITEM(pyinput, i, inputs[i].release());
+  }
+  return pyinput;
+}
+
+static variable_list unwrap_variable_list(PyObject* pyresult) {
+  TORCH_CHECK(PyList_CheckExact(pyresult));
+  auto result_len = PyList_GET_SIZE(pyresult);
+  variable_list result;
+  result.reserve(result_len);
+  for (const auto i : c10::irange(result_len)) {
+    result.emplace_back(THPVariable_Unpack(PyList_GET_ITEM(pyresult, i)));
+  }
+  return result;
+}
+static PyObject* check(PyObject* pyresult) {
+  if (C10_UNLIKELY(pyresult == nullptr)) {
+    // see https://github.com/pytorch/pytorch/pull/34845
+    python_error err;
+    err.persist();
+    throw err;
+  }
+  return pyresult;
+}
+
+static void check(bool result) {
+  if (C10_UNLIKELY(!result))
+    check(nullptr);
+}
+
+struct CacheNode {
+  static CacheNode* root() {
+    static CacheNode _root;
+    return &_root;
+  }
+
+  CacheNode* lookup(const CacheKey& key) {
+    auto it = next.find(key);
+    if (it == next.end()) {
+      // caller's key is in temporary memory, must copy it
+      CacheKeyBuffer buffer(key.key, key.key_size);
+      CacheKey key_with_storage(key.node_type, buffer.data, key.key_size);
+      it = next.emplace(key_with_storage, std::make_unique<CacheNode>()).first;
+      key_storage.emplace_back(std::move(buffer));
+    }
+    return it->second.get();
+  }
+
+  void clear() {
+    next.clear();
+    key_storage.clear();
+    expected_sizes.clear();
+    compiled_fn = nullptr;
+  }
+
+  bool is_empty() const {
+    return next.size() == 0 && !compiled_fn;
+  }
+
+  CacheNode() : compiled_fn(nullptr) {}
+  CacheNode(CacheNode&&) = delete;
+  CacheNode(const CacheNode&) = delete;
+  CacheNode& operator=(const CacheNode&) = delete;
+  CacheNode& operator=(CacheNode&&) = delete;
+
+  bool check_dynamic_sizes(AutogradCompilerCall& call) {
+    /*
+    We start off by assuming everything is static, then we mark things
+    as dynamic when we see them change.  This function:
+      1) Checks for a cache hit
+      2) Updates expected_sizes to track what is dynamic
+      3) Populates call.dyn_size_inputs by filtering call.all_size_inputs
+    */
```

Comment:
sounds familiar :) 

Note - dynamo has a plan to do automatic dynamic strides, do you want to point out that (a) we don't handle strides here either, and (b) there is an existing system in dynamo that mirrors this/this mirrors?

Reply:
Unfortunately we have no way of knowing what is a stride and what is a size at the level of autograd.  So we treat them all the same.